src/atcalls.c

```diff
@@ -94,22 +96,105 @@ _fullpathat(int dirfd, const char *relative, char *buf)
 
 int faccessat(int dirfd, const char *pathname, int mode, int flags)
 {
-    struct attrlist al = {
-        .bitmapcount = ATTR_BIT_MAP_COUNT,
-        .commonattr  = ATTR_CMN_USERACCESS,
-    };
-    struct {
-        uint32_t length;
-        uint32_t access;
-    } __attribute__((aligned(4), packed)) ab;
-    unsigned long opts = 0;
-
-    ERR_ON(EINVAL, flags & ~AT_SYMLINK_NOFOLLOW);
-    if (flags & AT_SYMLINK_NOFOLLOW)
-        opts |= FSOPT_NOFOLLOW;
-    if (getattrlistat(dirfd, pathname, &al, &ab, sizeof(ab), opts) < 0)
+    // Check that the arguments are valid
+
+    ERR_ON(EFAULT, pathname == NULL);
+    ERR_ON(EBADF, pathname[0] != '/' && dirfd < 0 && dirfd != AT_FDCWD);
+    ERR_ON(EINVAL, mode != F_OK && (mode & ~(R_OK|W_OK|X_OK)));
+    ERR_ON(EINVAL, flags & ~(AT_EACCESS|AT_SYMLINK_NOFOLLOW));
+
+    // First we stat the file.
+    // FIXME?: This doesn't properly support mode == F_OK && !(flags & AT_EACCESS).
+    // i.e. fstatat's own access will always be effective user access, not real user.
+    // This matters when the executability of an ancestor directory in a
+    // relative path is different for the real user and the effective user.
+    // But F_OK wasn't supported at all before, so presumably the impact is low.
+    // At least it's partially supported now. BUT THAT'S PROBABLY A SECURITY FLAW.
+    // ACTUALLY, THE SAME ISSUE PROBABLY APPLIES TO ALL USES IN SETUID/SETGID PROGRAMS!
+    // But faccessat is probably mostly used with a file name in the dirfd directory.
+    // After all, the whole point of faccessat is to avoid path-based race conditions.
+    // When that's the case, it behaves correctly.
+    // And it's probably no worse than before openat et al. existed(?).
+    // A complete solution would probably require iterating the following checks
+    // (adjusted appropriately) over all the elements of the relative pathname.
+    // Doing that in userspace while avoiding race conditions is a challenge.
+    // That's why it's a system call. Anyone up for it? It's probably not hard.
+    // One (generous) way to look at this implementation is that it provides
+    // the expected security benefits of faccessat when used with a pathname
+    // that doesn't contain any / characters. But, when used with more complex
+    // pathnames, it's no better (or worse(?)) than before faccessat existed.
+
+    struct stat st;
+    if (fstatat(dirfd, pathname, &st, flags & AT_SYMLINK_NOFOLLOW) < 0)
         return -1;
-    ERR_ON(EACCES, (ab.access & mode) != (uint32_t)mode);
+
+    // Check F_OK (fstatat above succeeded, so the file exists)
+
+    if (mode == F_OK)
+        return 0;
```

Comment:
This is based on assumption: if `stat` works => process has access to it and it's ok.

I not sure that it's true for `euid`/`egid` case.

Maybe move it after:
```
ERR_ON(EACCES, (st.st_mode & eperms) != eperms)
```
to ensure all checks?

Reply:
No. It's based on the assumption that if stat works, then the file exists. There is no assumption about the (`rwx`) access to the file. Note that `faccessat` can be called with `mode` equal to `F_OK` or a bitmask of the access bits. They can't be combined. So if `F_OK` is specified, there are no access checks to be performed.

However, as explained in the comment above that code, if the `pathname` contains any `/` characters, then it is always the effective users' permissions that determines access to the directories that lead to the file, not the real user's permissions. And although I do think that's incorrect, and that a real system call could do it better, I'm not sure that it's worth fixing that aspect. As I said in the comment, I think that it makes most sense to use `at` functions with `dirfd` referring to the directory that directly contains the file, and `pathname` would not normally have any '/' characters. When that's the case, it's the effective user's permissions that apply to navigating to the directory in question anyway. So I think that in most cases (i.e. when using `at` functions in order to avoid path-based race conditions) this behaviour is indistinguishable from correct behaviour.

Reply:
Actually, it doesn't make sense to knowingly leave it buggy in some cases. I was just being lazy. I'll make it work properly when `pathname` contains `/` characters.